budgies/threshold_optimizer.py

```diff
@@ -1,3 +1,251 @@
+        self.optimized_metrics = dict()
+        self._supported_metrics = [
+            'f1', 'accuracy', 'sensitivity', 'specificity',
+            'precision', 'recall',
+        ]
+
+    def set_search_space(self,
+                         search_space_size: int):
+        """set the number of possible probability threshold values to optimze for
+
+        This function is useful to reset the size of the search space after initializing the ThresholdOptimizer object.
+
+        Args:
+            search_space_size: The number of possible probability threshold values to optimze for
+        """
+        self.search_space = np.linspace(0, 1, search_space_size)
+
+    def convert_classes(self,
+                        threshold: int) -> np.ndarray:
+        """Convert predicted probabilities into binary classes based on a threshold/cutoff value
+
+        Args:
+            threshold: The probability threshold value to determine predicted classes.
+                        This follows a greater than or equal to format for determining class 1
+
+        Returns: 1 dimensional numpy array of classes
+
+        """
+        classes = np.where(self.predicted_probabilities >= threshold, 1, 0)
+        return classes
+
+    def _get_best_metrics(self,
+                          metric_type: str,
+                          scores: list,
+                          optimization: str = 'max',
+                          verbose: bool = True) -> Tuple[int, int]:
+        """computes optimized metrics based which supported metric was specified
+
+        Args:
+            metric_type: The name of the mertic to optimize for. It should be one of the supported metrics
+            scores: Computed metrics for all threshold values in the search space
+            optimization: Optional. Indicator of whether to optimize by finding the maximum metric value
+                            or the minimum metric value
+            verbose: Optional. Option of whether to output results of optimization. Defaults to true
+
+        Returns: Best score and best threshold for a specified metric
+
+        """
+        if optimization.lower() == 'max':
+            best_score = max(scores)
+        elif optimization.lower() == 'min':
+            best_score = min(scores)
+        else:
+            raise ValueError('Wrong value passed into optimization parameter. Should be max or min')
+        best_index = scores.index(best_score)
+        best_threshold = self.search_space[best_index]
+        self.optimized_metrics.update(
+            {
+                metric_type: {
+                    'best_score': best_score,
+                    'best_threshold': best_threshold,
+                    'all_scores': scores,
+                },
+            },
+        )
+        if verbose:
+            print(f'best {metric_type}: {best_score} occurs at threshold {best_threshold}')
+        return best_score, best_threshold
+
+    def get_best_f1_metrics(self,
```

Comment:
Agree. I kept it that way because we didn't ever get the time to discuss what we each did. I built it in such a way that anyone can take look at the function they worked on and maintain a high-level of readability. I'm aligned with you though.